pkg/geo/geomfn/subdivide.go

```diff
@@ -0,0 +1,364 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package geomfn
+
+import (
+	"math"
+
+	"github.com/cockroachdb/cockroach/pkg/geo"
+	"github.com/cockroachdb/errors"
+	"github.com/twpayne/go-geom"
+)
+
+// Subdivide returns a geometry divided into parts, where each part contains no more than the number of vertices provided.
+func Subdivide(g geo.Geometry, maxVertices int) (geo.Geometry, error) {
+	if g.Empty() {
+		return g, nil
+	}
+	const minMaxVertices = 5
+	if maxVertices < minMaxVertices {
+		return geo.Geometry{}, errors.Newf("max_vertices number cannot be less than %v", minMaxVertices)
+	}
+
+	gt, err := g.AsGeomT()
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform input geometry into geom.T: %v", err)
+	}
+	col := geom.NewGeometryCollection()
+	dim, err := dimensionFromGeomT(gt)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not calculate geometry dimension: %v", err)
+	}
+	const startDepth = 0
+	err = subdivideRecursive(gt, maxVertices, startDepth, col, dim)
+	if err != nil {
+		return geo.Geometry{}, err
+	}
+	col.SetSRID(gt.SRID())
+	output, err := geo.MakeGeometryFromGeomT(col)
+	if err != nil {
+		return geo.Geometry{}, errors.Newf("could not transform output geom.T into geometry: %v", err)
+	}
+	return output, nil
+}
+
+func subdivideRecursive(
+	gt geom.T, maxVertices int, depth int, col *geom.GeometryCollection, dim int,
+) error {
+	const geometryCollectionPushError = "could not add geometry to results collection: %v"
+	// maxDepth 50 => 2^50 ~= 10^15 subdivisions.
+	const maxDepth = 50
+	clip := geo.BoundingBoxFromGeomTGeometryType(gt)
+	width := clip.HiX - clip.LoX
+	height := clip.HiY - clip.LoY
+	if width == 0 && height == 0 {
+		if gt, ok := gt.(*geom.Point); ok && dim == 0 {
```

Comment:
when would dim not be 0 here?

Reply:
It would not be `dim == 0` if the input `geom.T` is a type of `geom.GeometryCollection` that includes a `Point` and any other type.
For example - if my collection consists of a `LineString` and a `Point`, the output collection should consist only of a `LineString`s subdivisions, while `Point` should be ignored because it has a lower dimension. However, since the dimension is checked later in the code, the `dim` value would be different than `0` here and the `Point` would be added to the output collection anyway. I have just added a test case to cover that.

That's a good question, I have not realized it earlier (I copied from the PostGIS algorithm, to be honest)